### ISSUE TITLE
Conversion traits should emphasize their intended use and limitations

### ISSUE THREAD

**Issue by Gankro** [rust-lang/rust#29701]
The conversion traits, in my books, serve one primary role: convenient generic one-time conversions to a concrete type to make APIs "just work". For instance, `fn foo<P: AsRef<Path>>(path: P)` "just works" for the various kinds of string.

Invoking them in this kind of generic context is pretty much guaranteed to work because of the magic of trait resolution in generic contexts. However outside of this context, they can and will behave in wild and uncontrolled ways. There are two major issues I see:

1) In concrete contexts, the pervasiveness of implementations of this trait means it will basically always do _something_, but deref coercions and inference mean that it's unlikely to be the thing you want, nor is it likely to do it in a stable manner. So for instance,

`let x = y.as_ref()` may plausibly yield `y`, `&y`, `&*y`,  `&**y`, or some particular concrete impl depending on the type of y (and x?).

2) You can't trust traits. Just because something implements AsRef and AsMut, doesn't mean they agree. Just because you called AsRef once before, doesn't mean the next call will yield the same result. These traits are for _conversions_ -- not for building some kind of generic polymorphism framework on top of.

**Comment by Gankro** [comment]
cc @aturon @SimonSapin

**Comment by SimonSapin** [comment]
Sounds fair enough.

**Comment by SimonSapin** [comment]
For context: [this library](https://github.com/SimonSapin/rust-std-candidates/blob/90b31dfafd7c6b6d15905db7a5db7836e6a6105f/string-wrapper/lib.rs) has a struct that holds a value of generic type `T` that represents a bytes buffer. The first version used `T: AsRef<[u8]> + AsMut<[u8]>` until @bluss pointed out that bad impls could violate assumptions that library makes in unsafe code. I changed the bound to a custom `unsafe` trait.

**Comment by steveklabnik** [comment]
part of https://github.com/rust-lang/rust/issues/29349

**Comment by steveklabnik** [comment]
@Gankro Which traits should be included in this audit?

* AsRef/AsMut
* From/Into

Anything else?

**Comment by Gankro** [comment]
Discussing Borrow/ToOwned might also be relevant? In particular when to use Borrow vs AsRef.

**Comment by Protagores** [comment]
There might be some traits that would just require consistency of the results of its method(s) to be deemed "trustworthy", and only the values of the arguments would be enough for calculating the results, neither any side effects need to occur. This looks suspiciously similar to limitations imposed on CTFE code - its results have to be consistent, it only has access to input arguments and can't cause side effects.

Basically, once CTFE is introduced in Rust (see rust-lang/rfcs#322), it might be reasonable to mark those methods in such traits as `const fn`, imposing the corresponding restrictions on its implementations. This is most likely going to break a few things in the ecosystem, however I'm not sure if the tradeoffs of this change would be really significant.

**Comment by steveklabnik** [comment]
It is unclear to me how relevant this issue is today. It's been open for a really long time with no comments. We've re-done `std::convert` in the meantime. If there's any specific traits that need better explanations, please open new issues with details of what needs clarified. Thanks!
